src/controllers/dlgprefcontroller.cpp

```diff
@@ -303,94 +368,115 @@ void DlgPrefController::slotUpdate() {
 }
 
 void DlgPrefController::slotCancel() {
-    if (m_pInputTableModel != NULL) {
-        m_pInputTableModel->cancel();
+    slotShowPreset(m_pController->getPreset());
+}
+
+void DlgPrefController::applyPresetChanges() {
+    if (m_pInputTableModel) {
+        m_pInputTableModel->apply();
     }
 
-    if (m_pOutputTableModel != NULL) {
-        m_pOutputTableModel->cancel();
+    if (m_pOutputTableModel) {
+        m_pOutputTableModel->apply();
     }
 }
 
 void DlgPrefController::slotApply() {
-    if (m_bDirty) {
-        // Apply the presets and load the resulting preset.
-        if (m_pInputTableModel != NULL) {
-            m_pInputTableModel->apply();
-        }
-
-        if (m_pOutputTableModel != NULL) {
-            m_pOutputTableModel->apply();
-        }
+    applyPresetChanges();
 
-        // Load the resulting preset (which has been mutated by the input/output
-        // table models). The controller clones the preset so we aren't touching
-        // the same preset.
-        emit loadPreset(m_pController, m_pPreset);
+    // If no changes were made, do nothing
+    if (!(isDirty() || (m_pPreset && m_pPreset->isDirty()))) {
+        return;
+    }
 
-        //Select the "..." item again in the combobox.
-        m_ui.comboBoxPreset->setCurrentIndex(0);
+    bool bEnabled = false;
+    if (m_pPreset) {
+        bEnabled = m_ui.chkEnabledDevice->isChecked();
 
-        bool wantEnabled = m_ui.chkEnabledDevice->isChecked();
-        bool enabled = m_pController->isOpen();
-        if (wantEnabled && !enabled) {
-            enableDevice();
-        } else if (!wantEnabled && enabled) {
-            disableDevice();
+        if (m_pPreset->isDirty()) {
+            savePreset();
         }
+    }
+    m_ui.chkEnabledDevice->setChecked(bEnabled);
 
-        m_bDirty = false;
+    // The shouldn't be dirty at this pint because we already tried to save
+    // it. If that failed, don't apply the preset.
+    if (m_pPreset && m_pPreset->isDirty()) {
+        return;
     }
+
+    // Load the resulting preset (which has been mutated by the input/output
+    // table models). The controller clones the preset so we aren't touching
+    // the same preset.
+    emit applyPreset(m_pController, m_pPreset, bEnabled);
+
+    // Mark the dialog as not dirty
+    setDirty(false);
 }
 
-void DlgPrefController::slotLoadPreset(int chosenIndex) {
+void DlgPrefController::slotPresetSelected(int chosenIndex) {
+    QString presetPath;
     if (chosenIndex == 0) {
-        // User picked ...
-        return;
+        // User picked "No Preset" item
+        m_ui.chkEnabledDevice->setEnabled(false);
+    } else {
+        // User picked a preset
+        m_ui.chkEnabledDevice->setEnabled(true);
```

Comment:
add
`m_ui.chkEnabledDevice->setChecked(true);`

Reply:
Hmm, so that selecting a preset means activating the controller? In that case we can remove the enabled checkbox completely and just use "no preset" for disabled controllers.

Reply:
Is there a use case for keeping a mapping selected but disabling the controller? The only time I have needed to disable and re-enable a controller was when the USB cable came unplugged during a set.

Reply:
Wait, this works? I always restarted mixxx in that case.

Reply:
Indeed, it does. It's not obvious at all though.

Reply:
You have to manually disable the checkbox then re-enable it

Reply:
As a stopgap until we support controller hotplug, how about getting rid of the Enabled checkbox and treating every controller with a selected preset as enabled. For the reconnection edge case, we can add a Reconnect button that does what unchecking and rechecking the Enabled checkbox does now.

Reply:
Hmm, how about we leave it as-is for 2.3? Users that upgrade from 2.2 will have a preset defined for all controllers because Mixxx was writing custom presets everywhere. This means that all controllers would be active. During development I made a little mistake that also activated all controllers - this made both my keyboard and my mouse unresponsive since they are also HID devices. I had to restart X11 via SSH to make my computer usable again :-)

Reply:
Oof, yeah. I was wondering if those legacy files laying in the user folders would have consequences. They sure would if we automatically enabled them. I still think it would help to add `m_ui.chkEnabledDevice->setChecked(true);` here in slotPresetSelected so users don't have to think about that checkbox.

Reply:
I think we can do this without automatically activating those legacy presets in the user folder. Here is what a fresh mixxx.cfg looks like from master after activating one of my Xone K2s:
```
Babyface_Pro_(70785713)_MIDI_1 /home/be/.mixxx/controllers/Babyface_Pro_(70785713)_MIDI_1.midi.xml
Babyface_Pro_(70785713)_MIDI_2 /home/be/.mixxx/controllers/Babyface_Pro_(70785713)_MIDI_2.midi.xml
XONE:K2_MIDI_1 /home/be/.mixxx/controllers/Allen and Heath Xone K2.midi.xml

[Controller]
XONE:K2_MIDI_1 1
```
I propose still relying on the `[Controller]` ConfigKey to determine which controllers to enable. Then automatically enable that old ConfigKey when a preset is applied. This would allow us to get rid of the Enabled checkbox.

Reply:
Ok, this sounds like it could work. However, let's do this in a separate PR, this one is already big enough and I have the impression that PRs always take a few days without any new commits before they get merged here :wink:

Reply:
Okay